mitodl/ol-django
# Adds transaction search and lookup to the CyberSource payment gateway

#### Pre-Flight checklist


#### What are the relevant tickets?

Partially fixes https://github.com/mitodl/mitxonline/issues/1415 

#### What's this PR do?

Adds transaction search and detail lookup to the CyberSource payment gateway class. 

* `find_transactions` will create and perform a search using the CyberSource API and return transaction IDs that match the supplied app-specific order IDs.
* `get_transaction_details` will return data about the supplied CyberSource transaction ID. This will return both the [CyberSource API-specific value](https://github.com/CyberSource/cybersource-rest-client-python/blob/master/docs/TssV2TransactionsGet200Response.md) for this and that data reformatted to fit the normal transaction payload that you'd get back from creating a payment.
* `find_and_get_transactions` combines the two above calls, so you can pass it a list of order IDs and get the transaction details for each, with some resolution in case there are multiple transactions for the same order ID. (This is the call that I expect will be used most often.)

#### How should this be manually tested?

Automated tests should pass. 

Since this code is mostly for MITx Online, the easiest way to test this is within that system:
* Make sure your MITx Online instance has a full set of CyberSource credentials, _including the REST API keys_. You should ideally have some completed transactions as well, and it wouldn't hurt (but is not necessary) to be logged into the test EBC as well to verify things.
* Build a stand-alone version of the mitol-django-payment_gateway package: `./pants package "src/mitol/payment_gateway:"` This will result in a zip file in the `dist` folder with the pacakge in it. 
* Copy the resulting zip file into your MITx Online instance and expand it somewhere. 
* Install the package: `pip install -e <path to the package folder>` (In my case, it ended up being just `pip install -e payment_gateway/mitol-django-payment-gateway-1.8.1` as I'd put it in a subfolder.) 
* Test running the methods. As there won't be any code in MITx Online that uses this for now, you can test via a Django shell or Jupyter notebook:

```python
from mitol.payment_gateway.api import CyberSourcePaymentGateway

cspg = CyberSourcePaymentGateway()
# get some order IDs in whatever way you see fit, put them in a list in orders
search_results = cspg.find_transactions(orders)
transaction_details = cspg.find_and_get_transactions(orders)
# inspecting search_results and transaction_details should return data that makes sense
# you can also do:
cs_transaction_details = cspg.get_transaction_details(transaction_id)
# where transaction_id is one of the CyberSource IDs; you should get back a tuple.
```
